Protect RPC interface against CSRF

kingosticks opened the pull request:
A fix for some of https://github.com/mopidy/mopidy/issues/1659

kingosticks commented:
I will take a stab at the websocket too

jodal commented on mopidy/http/ext.conf:
```diff
@@ -4,3 +4,4 @@ hostname = 127.0.0.1
 port = 6680
 static_dir =
 zeroconf = Mopidy HTTP server on $hostname
+allowed_origins =
```
Does `localhost:6600` work without whitelisting it, or should we make that the default?

jodal commented on docs/ext/http.rst:
```diff
@@ -98,3 +98,11 @@ See :ref:`config` for general help on configuring Mopidy.
+.. confval:: http/allowed_origins
+
+    A whitelist of domains allowed to perform Cross-Origin Resource Sharing
+    (CORS) requests. Entries must be in the format ``hostname``:``port``.
```
- [ ] Consider s/whitelist/list/ and explain that multiple values can be separated by comma or put on separate lines. I don't remember how well we explain how to use `config.List` config values elsewhere.
- [ ] Include the colon in the monospace part: ``hostname:port``.

jodal commented on mopidy/http/handlers.py:
```diff
@@ -19,12 +20,16 @@
 
 def make_mopidy_app_factory(apps, statics):
     def mopidy_app_factory(config, core):
+        origin_whitelist = {
```
s/origin_whitelist/allowed_origins/ to keep the naming consistent?

jodal commented on mopidy/http/handlers.py:
```diff
@@ -143,16 +148,31 @@ def set_mopidy_headers(request_handler):
+    if origin is None:
+        logger.debug('Origin was not set')
+        return False
+    origin_whitelist.add(request_headers.get('Host', None))
```
This allows access for a webpage at `foo:80` if the RPC interface also is available at `foo:80`? Cute. That voids my comment about `localhost:6600` as the default config value.

jodal commented on mopidy/http/handlers.py:
```diff
@@ -177,6 +197,18 @@ def set_extra_headers(self):
         self.set_header('Content-Type', 'application/json; utf-8')
 
+    def options(self):
+        origin = self.request.headers.get('Origin', None)
```
`None` is the default return value for `.get()`, so you can leave it out.

kingosticks commented:
So I think I have addressed those comments.

Now for the websocket, I thought we could just use the same function:
```
    def check_origin(self, origin):
        return check_origin(origin, self.request.headers, self.allowed_origins)
```

kingosticks commented:
Fixed conflict and added a bit about same-origin requests to the docs and a changelog entry.

jodal commented:
Thanks for taking the time to fixing this properly :-)
